common/schemas/keyman-touch-layout/keyman-touch-layout.clean.spec.json

```diff
@@ -0,0 +1,154 @@
+{
+  "$schema": "http://json-schema.org/schema#",
+  "$ref": "#/definitions/touch-layout",
+  "$comment": "Version: 16.0",
+  "description": "A Keyman Touch Layout file, per version 16.0, clean spec, no legacy data or types",
+
+  "definitions": {
+    "touch-layout": {
+      "type": "object",
+      "properties": {
+        "tablet": { "$ref": "#/definitions/platform" },
+        "phone": { "$ref": "#/definitions/platform" },
+        "desktop": { "$ref": "#/definitions/platform" }
+      }
+    },
+
+    "platform": {
+      "type": "object",
+      "properties": {
+        "font": { "$ref": "#/definitions/font-spec" },
+        "fontsize": { "$ref": "#/definitions/fontsize-spec" },
+        "layer": { "$ref": "#/definitions/layers" },
+        "displayUnderlying": { "type": "boolean" }
+      },
+      "required": ["layer"],
+      "additionalProperties": false
+    },
+
+    "layers": {
+      "type": "array",
+      "items": { "$ref": "#/definitions/layer" }
+    },
+
+    "layer": {
+      "type": "object",
+      "properties": {
+        "id": { "$ref": "#/definitions/layer-id" },
+        "row": { "$ref": "#/definitions/rows" }
+      },
+      "required": ["id", "row"],
+      "additionalProperties": false
+    },
+
+    "layer-id": {
+      "type": "string",
+      "pattern": "^[a-zA-Z0-9_-]+$"
+    },
+
+    "rows": {
+      "type": "array",
+      "items": { "$ref": "#/definitions/row" }
+    },
+
+    "row": {
+      "type": "object",
+      "properties": {
+        "id": { "$ref": "#/definitions/row-id" },
+        "key": { "$ref": "#/definitions/keys" }
+      },
+      "required": ["id", "key"],
+      "additionalProperties": false
+    },
+
+    "row-id": {
+      "type": "integer",
+      "minimum": 0,
+      "maximum": 100
+    },
+
+    "keys": {
+      "type": "array",
+      "items": { "$ref": "#/definitions/key" }
+    },
+
+    "key": {
+      "type": "object",
+      "properties": {
+        "id": { "$ref": "#/definitions/key-id" },
+        "text": { "type": "string" },
+        "layer": { "$ref": "#/definitions/layer-id" },
+        "nextlayer": { "$ref": "#/definitions/layer-id" },
+        "font": { "$ref": "#/definitions/font-spec" },
+        "fontsize": { "type": "string" },
+        "sp": { "$ref": "#/definitions/key-sp" },
+        "pad": { "$ref" : "#/definitions/key-pad" },
+        "width": { "$ref" : "#/definitions/key-width" },
+        "sk": { "$ref": "#/definitions/subkeys" },
+        "flick": { "$ref": "#/definitions/flick" },
+        "multitap": { "$ref": "#/definitions/subkeys" }
+      },
+      "additionalProperties": false
+    },
+
+    "key-id": {
+      "type": "string",
+      "pattern": "^[TKUtku]_[a-zA-Z0-9_]+$"
+    },
+
+    "key-sp": {
+      "type": "integer",
+      "enum": [0, 1, 2, 8, 9, 10],
```

Comment:
3 and 4 are also supported.  Those are for use of a keyboard-specific special OSK font for the key caps, I believe.

Reply:
Curious -- they are not listed anywhere in Keyman Developer:

https://github.com/keymanapp/keyman/blob/ec690f47f9979a9e8144dd3a9f1b06a4c820096f/developer/src/tike/xml/layoutbuilder/builder.xsl#L84-L91

By the looks of this, it seems that 3 and 4 are reserved by KeymanWeb as a rewiring of 1 & 2:

https://github.com/keymanapp/keyman/blob/4f36c3c2df0b14ec2d5fac36714c4e8bda54961c/common/web/keyboard-processor/src/keyboards/activeLayout.ts#L292-L306

I also see that KMW is treating `sp` as a string whereas this spec is forcing it to an integer. Given that we need to make sure that this is backwardly compatible for compiled keyboards, do we need a "legacy.spec.json" that covers this difference off along with any other similar fields (`pad`, `width`, I am looking at you)?

I don't mind the source format diverging from the compiled format; there are various things we can do to optimise the compiled keyboards if we go this way.

Reply:
I've made this mistake before; there _are_ keyboards where it matters.  Please refer to #4282.

Reply:
The keyboard itself isn't using `"sp": "3"` or `"sp": "4"`. Those are values assigned internally within KeymanWeb. Thus, I don't think we need to support those values in the schema.